File: backend/magnet_api/app/schemas/sync.py
from typing import List, Optional, Literal
from pydantic import BaseModel

class Metadata(BaseModel):
    filename: str
    size: int
    timestamp: int
    hash: str
    type: Literal['download', 'magnet']

class Change(BaseModel):
    id: str
    data: Optional[str] = None  # Base64 encoded data
    metadata: Metadata
    operation: str  # 'create', 'update', or 'delete'
    timestamp: int

class SyncPushRequest(BaseModel):
    changes: List[Change]
    lastSyncTimestamp: int
    deviceId: str

class SyncPullResponse(BaseModel):
    changes: List[Change]
    timestamp: int

File: frontend/src/types/sync.ts
export interface SyncMetadata {
  filename: string;
  size: number;
  timestamp: number;
  type: 'download' | 'magnet';
  isAdmin?: boolean;
  syncStatus: 'synced' | 'pending' | 'conflict';
  lastSynced: string | null;
  syncVersion: number;
  localChanges?: any;
  syncMetadata?: any;
}

export interface SyncChange {
  id: string;
  data: string | null;
  metadata: SyncMetadata;
  operation: 'create' | 'update' | 'delete';
  timestamp: number;
}

export interface SyncPushRequest {
  changes: SyncChange[];
  lastSyncTimestamp: number;
  deviceId: string;
}

export interface SyncPullResponse {
  data: {
    changes: SyncChange[];
    timestamp: number;
  };
}

export interface SyncState {
  lastSyncTimestamp: number;
  isSyncing: boolean;
  error: string | null;
}
